src/Code/Diagnostics/Command.cs

```diff
@@ -0,0 +1,152 @@
+using System;
+using System.Diagnostics;
+using System.Runtime.InteropServices;
+using System.Text;
+using System.Threading;
+using Microsoft.Extensions.Logging;
+
+namespace Code.Diagnostics
+{
+	enum CommandType
+	{
+		Standard,
+		Terminal,
+		Shell
+	}
+
+	static class Command
+	{
+		/// <summary>
+		/// Invokes the given <paramref name="command"/> with the given <paramref name="arguments"/>,
+		/// using the given <paramref name="timeout"/> and <paramref name="logger"/>.
+		/// </summary>
+		/// <param name="command">The command to invoke.</param>
+		/// <param name="arguments">The arguments to pass to the invocation.</param>
+		/// <param name="timeout">The <see cref="TimeSpan"/> for how long to wait for the process to exit.</param>
+		/// <param name="logger">A logger for logging output and errors.</param>
+		/// <param name="commandType">The type of command.</param>
+		/// <param name="hasError">A callback for determining whether there are any errors in the output.</param>
+		internal static void Run(
+			string command,
+			string arguments,
+			TimeSpan timeout,
+			ILogger logger,
+			CommandType commandType = CommandType.Standard,
+			Func<string, bool> hasError = null)
+		{
+			if (command == null) throw new ArgumentNullException(nameof(command));
+			if (arguments == null) throw new ArgumentNullException(nameof(arguments));
+			if (timeout == TimeSpan.Zero) throw new ArgumentException("Timeout must be non-zero.", nameof(timeout));
+			if (logger == null) throw new ArgumentNullException(nameof(logger));
+
+			using var process = StartProcess(command, arguments, commandType, logger, out var output, out var error);
+
+			if (commandType == CommandType.Terminal)
+			{
+				var exitCode = RuntimeInformation.IsOSPlatform(OSPlatform.Windows) ? "%errorlevel%" : "$?";
+
+				process.StandardInput.WriteLine($"{command} {arguments}");
+				process.StandardInput.WriteLine($"exit {exitCode}");
+			}
+
+			var milliseconds = Math.Abs((int)timeout.TotalMilliseconds);
+			if (process.WaitForExit(milliseconds) &&
+				output.WaitHandle.WaitOne(milliseconds) &&
+				error.WaitHandle.WaitOne(milliseconds))
+			{
+				hasError ??= text => false;

```

Comment:
You might want to be careful about using newer language features, unless you have a mechanism in msbuild to check the language version, this is difficult though as there is no msbuild way of resolving the actual language version, it's all passed into csc to figure out.

Reply:
That's a very good point that I hadn't considered. Well spotted!